File: 62-UniquePaths/62-UniquePaths.cpp
// Last updated: 04/04/2026, 23:38:01
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> chache(m,vector<int>(n,0));
        for(int i=0;i<m;i++) chache[i][0] = 1;
        for(int i=0;i<n;i++) chache[0][i] = 1;
        for(int i=1;i<m;i++) {
            for(int j=1;j<n;j++) {
                chache[i][j] = chache[i-1][j] + chache[i][j-1];
            }
        }
        return chache[m-1][n-1];
    }
};

File: 62-UniquePaths/62-UniquePaths.java
// Last updated: 04/04/2026, 23:38:05
class Solution {
    public static int m;
    public static int n;
    public int[][] mem;
    public int help(int row,int col) {
        if(row<0 || col<0) return 0;
        if(row==0 && col==0) {
            return 1;
        }
        if(mem[row][col]!=0) return mem[row][col];
        int  fromLeft = help(row-1,col);
        int fromTop = help(row,col-1);
        mem[row][col] = fromLeft + fromTop;
        return mem[row][col];
    }
    public int uniquePaths(int m, int n) {
        this.m = m;
        this.n = n;
        this.mem = new int[m][n];
        return help(m-1,n-1);
    }
}